lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,256 @@
+import 'dart:collection';
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+Element _getLeftElement(AssignmentExpression assignment) {
+  final leftPart = assignment.leftHandSide;
+  return leftPart is SimpleIdentifier
+      ? DartTypeUtilities.getCanonicalElement(leftPart.bestElement)
+      : leftPart is PropertyAccess
+          ? DartTypeUtilities
+              .getCanonicalElement(leftPart.propertyName.bestElement)
+          : null;
+}
+
+bool _isBreakStatement(Statement statement) => (statement is BreakStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is BreakStatement));
+
+bool _isContinueStatement(
+        Statement statement) =>
+    (statement is ContinueStatement ||
+        (statement is Block &&
+            statement.statements.length == 1 &&
+            statement.statements.first is ContinueStatement));
+
+bool _isReturnStatement(Statement statement) => (statement is ReturnStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is ReturnStatement));
+
+abstract class ConditionScopeVisitor extends RecursiveAstVisitor {
+  final Queue<Queue<_ExpressionBox>> environments = new Queue();
```

Comment:
This is one of the few issues that I have not solved, what do you think is a better option here @bwilkerson?